### ISSUE TITLE
Schedule Parallelization

### ISSUE THREAD

**Issue by xentripetal** [andreakarasho/TinyEcs#24]
It would be great to eventually support automatically parallelizing systems in schedules based on their resource access.

I started work on porting Bevy's ECS scheduler to C# but put it on hold for now. I plan to come back to in the future, but if you're interested my draft state is here https://github.com/xentripetal/PolyGame/tree/73b5596a2ecb85016f9edc937b091c7ce05c1117/PolyECS/Scheduling

**Comment by andreakarasho** [comment]
I don't know how bevy manages multithreading systems, but i made something working here https://github.com/andreakarasho/TinyEcs/tree/feat/scheduler-multithread
The logic should [didn't test so much lol] split all systems which are not using borrowed/already-accessed resources.

In addition, systems are spawned with the `ThreadingType.Auto` as default value. You can change that ofc and force the running behaviour.

```csharp
scheduler.AddSystem(() => Console.WriteLine("Auto"), threadingType: ThreadingType.Auto);
scheduler.AddSystem(() => Console.WriteLine("Single"), threadingType: ThreadingType.Single);
scheduler.AddSystem(() => Console.WriteLine("Multi"), threadingType: ThreadingType.Multi);
```

**Comment by xentripetal** [comment]
Interesting strategy, thanks for the quick solution!

Bevy looks at the tables each system will read or write, run systems in parallel based on it,  and automatically inserts sync points/end defers between write-reads.

**Comment by andreakarasho** [comment]
merged https://github.com/andreakarasho/TinyEcs/pull/25
